CoreFoundation/PlugIn.subproj/CFBundle.c

```diff
@@ -714,7 +714,7 @@ static CFBundleRef _CFBundleCreate(CFAllocatorRef allocator, CFURLRef bundleURL,
 
     bundle->_url = newURL;
     
-#if !DEPLOYMENT_RUNTIME_OBJC && !DEPLOYMENT_TARGET_WINDOWS
+#if !DEPLOYMENT_RUNTIME_OBJC && !DEPLOYMENT_TARGET_WINDOWS && !DEPLOYMENT_TARGET_ANDROID
```

Comment:
I’ve gone through the code again and gated anything I could find about FHS out for Android.

Reply:
There seems to be a bug in github where I can’t reply directly to your next comment about path(toResource:ofType:) so I’ll add a note here. One of the goals of the Android port is that code for iOS should run on Android and it’s likely that code will use the path(toResource:ofType:) pattern somewhere. On Android It’s up to the Java side of the app to copy resources required into the cache/tmp directory then the Swift/Foundation side can find these resources inside this directory as if it was a bundle so the code runs unmodified.

Reply:
I appreciate the use case, but again, it breaks our contract that Bundle.main be where the resources for a specific bundle _actually are_ (not _potentially will be if the author is careful_), and that this be a directory that's either specific to the application or at worst contains existing application executable and resource data for that one application.

I outlined a workaround above — since apps need to write custom code to copy resources to some filesystem directory, it is no stretch, if they want `Bundle` API to fetch their own resources, to ask that they construct a `Bundle` instance to the copied data.

I will also reiterate that any library code using `Bundle(for:)` that builds with the Swift Package Manager and then asserts the returned bundle is exclusive to that library is essentially in undefined behavior territory, and should be fixed upstream.

Reply:
(Also — that breaks if multiple components on Android are using Swift, have equally-named resources or resource directories {which isn't that unlikely: I assume careless copying code could easily clobber an en.lproj folder}, and are running at the same time.)

Reply:
Perhaps there just is a misunderstanding here. This is not a system shared directory. It is the app's home, combination Documents and tmp directory on iOS with a path something like: /data/data/com.example.user.myapplication/cache and is the only directory an app has read write access to I am aware of. There's no candidate for a bundle directory other than the approach I suggest of copying resources on Android as resources are not unpacked from their .apk (zip file) on demand. This is what the zip file for an Android app looks like:

```
Users-iMac:swift-corelibs-foundation user$ unzip -l ~/swift-android-kotlin/app/build/outputs/apk/app-debug.apk
Archive:  /Users/user/swift-android-kotlin/app/build/outputs/apk/app-debug.apk
  Length      Date    Time    Name
---------  ---------- -----   ----
     2788  00-00-1980 00:00   AndroidManifest.xml
     1211  00-00-1980 00:00   META-INF/CERT.RSA
    31429  00-00-1980 00:00   META-INF/CERT.SF
    31367  00-00-1980 00:00   META-INF/MANIFEST.MF
      246  01-01-1970 01:00   META-INF/kotlin-runtime.kotlin_module
       43  01-01-1970 01:00   META-INF/kotlin-stdlib-jre7.kotlin_module
     2090  01-01-1970 01:00   META-INF/kotlin-stdlib.kotlin_module
    46044  00-00-1980 00:00   classes.dex
      752  00-00-1980 00:00   classes2.dex
      926  01-01-1970 01:00   kotlin/annotation/annotation.kotlin_builtins
     3689  01-01-1970 01:00   kotlin/collections/collections.kotlin_builtins
      726  01-01-1970 01:00   kotlin/internal/internal.kotlin_builtins
    14202  01-01-1970 01:00   kotlin/kotlin.kotlin_builtins
     2296  01-01-1970 01:00   kotlin/ranges/ranges.kotlin_builtins
     4866  01-01-1970 01:00   kotlin/reflect/reflect.kotlin_builtins
 10933048  00-00-1980 00:00   lib/armeabi-v7a/libFoundation.so
   359812  00-00-1980 00:00   lib/armeabi-v7a/libXCTest.so
   661056  00-00-1980 00:00   lib/armeabi-v7a/libc++_shared.so
  2009280  00-00-1980 00:00   lib/armeabi-v7a/libcurl.so
   668356  00-00-1980 00:00   lib/armeabi-v7a/libdispatch.so
 25911876  00-00-1980 00:00   lib/armeabi-v7a/libscudata.so
  2045828  00-00-1980 00:00   lib/armeabi-v7a/libscui18n.so
  1418788  00-00-1980 00:00   lib/armeabi-v7a/libscuuc.so
  1079184  00-00-1980 00:00   lib/armeabi-v7a/libsqlite3.so
  5259060  00-00-1980 00:00   lib/armeabi-v7a/libswiftCore.so
    31020  00-00-1980 00:00   lib/armeabi-v7a/libswiftGlibc.so
   506816  00-00-1980 00:00   lib/armeabi-v7a/libswiftRemoteMirror.so
   383140  00-00-1980 00:00   lib/armeabi-v7a/libswiftSwiftOnoneSupport.so
  2810068  00-00-1980 00:00   lib/armeabi-v7a/libswifthello.so
  1783196  00-00-1980 00:00   lib/armeabi-v7a/libxml2.so
      620  00-00-1980 00:00   res/anim-v21/design_bottom_sheet_slide_in.xml
      620  00-00-1980 00:00   res/anim-v21/design_bottom_sheet_slide_out.xml
...
```
What would you have me do? Mark the whole Bundle api unavailable? There is also no prospect of a OSS fix upstream to Alamofire as the code I’m using is a fork of a fork. If it were me I would make the fix I suggest apply to all Linux systems in the interests of compatibility. If I were to fix Alamofire there just be some other library which falls at this hurdle.

But we’ve discussed these issues a few times and we’re going around in circles. What is the minimum changes I have proposed you would have me remove to get this PR merged? I’ll have to maintain a fork (undesirable as it may be) to get what I need to get done done.

Reply:
The following changes are, I believe, at issue:

 - `Bundle(for:)` should not return an arbitrary bundle. It should stay callable. (My personal intent is to implement it on Linux as soon as resources allow.)

 - `Bundle.main` should not pick an arbitrary path whose forming substantially differs from other platforms. It should go on the existing fallback path unless that breaks existing _well-formed_ code.

 - `CFPreferences…` should not assume that any `CFBundle` path is writable or a correct location for user data. All paths returned by `CFBundle` must be treated as read-only.

Reply:
Also, someone should send Alamofire a patch, I believe.

Reply:
OK.. how does it look now?